conans/client/rest/uploader_downloader.py

```diff
@@ -256,7 +256,12 @@ def print_progress(output, units, progress=""):
 def call_with_retry(out, retry, retry_wait, method, *args, **kwargs):
     for counter in range(retry):
         try:
-            return method(*args, **kwargs)
+            response = method(*args, **kwargs)
```

Comment:
I am not sure if this would be the solution. Inspecting the code in RestClient I see:

```python
if not response.ok:
     output.error("\nError uploading file: %s, '%s'" % (filename, response.content))
     failed.append(filename)
else:
     pass
...
if failed:
     raise ConanException("Execute upload again to retry upload the failed files: %s"
                                 % ", ".join(failed))
```

So if the response is not ok, it would have raised an Exception. I am not sure what I am missing, I have read the thread in https://github.com/conan-io/conan/issues/4907, but not fully clear.

Reply:
The problem is that: `Uploader::upload()` calls `call_with_retry()`.
This one try to call the "_upload_file"-method multiple times, when the response from "_upload_file" is a `ConanException`. 
When there is a 500 status coming back inside `_upload_file`, the response is not an `ConanException`, so that the response inside `call_with_retry()` goes through without repeating. 

Then the result of this is first checked in RestV2Methods::_upload_files() :
```python
            try:
                response = uploader.upload(resource_url, files[filename], auth=self.auth,
                                           dedup=self._checksum_deploy, retry=retry,
                                           retry_wait=retry_wait,
                                           headers=self._put_headers)
                self._output.writeln("")
                if not response.ok:
                    self._output.error("\nError uploading file: %s, '%s'" % (filename,
                                                                             response.content))
                    failed.append(filename)
                else:
                    pass
```
But at this stage there is no retry anymore possible.

I think one possibility is to check just for status 500 and to throw an ConanException inside `_upload_file()` or to check if the response is `ok`. That would be the way I have done it (also I don't like it, because download & upload is done in the some function (therefor I have to check, if the function `ok` is available)

Reply:
I updated the pull request and added some tests. Please review.